sdk/resourcemanager/Azure.ResourceManager/src/GenericOperationSource.cs

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#nullable disable
+
+using System;
+using System.ClientModel.Primitives;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+
+namespace Azure.ResourceManager
+{
+    internal class GenericOperationSource<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)] T> : IOperationSource<T>
+    {
+        private readonly ArmClient _client;
+        private readonly bool _isResource;
+
+        public GenericOperationSource(ArmClient client, bool isResource)
+        {
+            _client = client;
+            _isResource = isResource;
+        }
+
+        T IOperationSource<T>.CreateResult(Response response, CancellationToken cancellationToken)
+            => CreateResult(response);
+
+        ValueTask<T> IOperationSource<T>.CreateResultAsync(Response response, CancellationToken cancellationToken)
+            => new ValueTask<T>(CreateResult(response));
+
+        private T CreateResult(Response response)
+        {
+            object data = ModelReaderWriter.Read(response.Content, typeof(T));
+            return _isResource
+                ? (T)Activator.CreateInstance(typeof(T), BindingFlags.NonPublic | BindingFlags.Instance, null, new object[] { _client, data }, null)
```

Comment:
is this working when T is a resource?
For example, if T is `VirtualMachineResource` here, therefore we are calling `ModelReaderWriter.Read` here on `VirtualMachineResource` type, what type is this method actually returning? Is it `VirtualMachineResource` or `VirtualMachineData`?
if it is `VirtualMachineData`, I believe everything should be fine - but if it is returning `VirtualMachineResource`, this CreateInstance call will not work because this ctor is expecting `VirtualMachineData` for its second parameter.

Reply:
well I think I just need which generic parameter of `IJsonModel<T>` on the resource is - I believe it should be `VirtualMachineData`, it looks like:
```
public partial class VirtualMachineResource: ArmResource, IJsonModel<VirtualMachineData>
```
Is this the case? If so, my concern above does not exist.

Reply:
If we take a look at https://github.com/Azure/azure-sdk-for-net/blob/416ea7dcc552bb26b72c27e8b5c89e9ab1fe0edf/sdk/resourcemanager/Azure.ResourceManager/src/Resources/Generated/ResourceGroupResource.Serialization.cs#L14-L18
`ResourceGroupResource` implements `IJsonModel<ResourceGroupData>`, so what `ModelReaderWriter.Read` returns is `object` but the actual content is `ResourceGroupData`.